### PR TITLE
ENH: Use Qt backend for doc build

### PR DESCRIPTION
**Pull request by larsoner**
@marsipu I can see the gc failure with the following in the `doc` folder:
```
$ make clean && PATTERN=15_inplace MNE_SKIP_INSTANCE_ASSERTIONS="brain,_Renderer,Plotter,BackgroundPlotter,vtkPolyData" make html_dev-pattern-memory
loading intersphinx inventory from https://www.fatiando.org/pooch/latest/objects.inv...
intersphinx inventory has moved: https://dipy.org/documentation/latest/objects.inv/ -> https://dipy.org/documentation/1.4.1./objects.inv/
generating gallery...
Using Sphinx-Gallery to convert rst text blocks to markdown for .ipynb files.
Clean    8.8 s : 442.5 MBto_tutorials/intro... [ 28%] 15_inplace.py                                                                                                                 

Traceback (most recent call last):
  File "/home/larsoner/python/sphinx/sphinx/events.py", line 101, in emit
    results.append(listener.handler(self.app, *args))
  File "/home/larsoner/python/sphinx-gallery/sphinx_gallery/gen_gallery.py", line 464, in generate_gallery_rst
    generate_dir_rst(src_dir, target_dir, gallery_conf,
  File "/home/larsoner/python/sphinx-gallery/sphinx_gallery/gen_rst.py", line 399, in generate_dir_rst
    intro, title, cost = generate_file_rst(
  File "/home/larsoner/python/sphinx-gallery/sphinx_gallery/gen_rst.py", line 1030, in generate_file_rst
    clean_modules(gallery_conf, fname, 'after')
  File "/home/larsoner/python/sphinx-gallery/sphinx_gallery/scrapers.py", line 596, in clean_modules
    reset_module(gallery_conf, fname, when=when)
  File "/home/larsoner/python/mne-python/doc/conf.py", line 346, in __call__
    _assert_no_instances(PyQtGraphBrowser, when)
  File "/home/larsoner/python/mne-python/mne/utils/misc.py", line 368, in _assert_no_instances
    assert n == 0, f'\n{n} {cls.__name__} @ {when}:\n' + '\n'.join(ref)
AssertionError: 
1 PyQtGraphBrowser @ mne/conf.py:Resetter.__call__:after:15_inplace.py:
dict: len=16, 8 referrers: dict/frame/list/module
dict: len=2, 4 referrers: LoadThread/dict/list
method: <bound method PyQtGraphBrowser.change_duration of <mne_qt_browser._pg_figure.PyQtGraphBrowser object
method: <bound method PyQtGraphBrowser.change_duration of <mne_qt_browser._pg_figure.PyQtGraphBrowser object
method: <bound method PyQtGraphBrowser.change_nchan of <mne_qt_browser._pg_figure.PyQtGraphBrowser object at
method: <bound method PyQtGraphBrowser.change_nchan of <mne_qt_browser._pg_figure.PyQtGraphBrowser object at
method: <bound method PyQtGraphBrowser.scale_all of <mne_qt_browser._pg_figure.PyQtGraphBrowser object at 0x
method: <bound method PyQtGraphBrowser.scale_all of <mne_qt_browser._pg_figure.PyQtGraphBrowser object at 0x
dict: len=5, 4 referrers: AnnotationDock/dict/list
...
```
It gives us at least some debugging leads on what could be causing `gc` not to be able to collect the `PyQtGraphBrowser` instances.

### CODE REVIEW

**Comment by marsipu** [comment]
Oh, this seems complicated. There is no immediate idea I get when looking at this.
PyQtBrowser has references of all items either stored as attributes or as attributes of `BrowserParams` (self.mne). Actually I never asked you what the initial design idea of this `BrowserParams` as `self.mne` was. Sometimes I set attributes like `LoadThread` directly because I thougt `self.mne` is only attributes common to `pyqtgraph` and `matplotlib`.

I thought, when a PyQtBrowser-object is garbage-collected all attributes should be garbace-collected too. Or what does this mean with the methods above?

**Comment by larsoner** [comment]
> I thought, when a PyQtBrowser-object is garbage-collected all attributes should be garbace-collected too. Or what does this mean with the methods above?

Yes in theory this should happen, the python garbage collector *should* recognize when all references are gone, or when the only remaining ones are circular. But once you start interacting with C++ libraries it becomes harder. @GuillaumeFavelier for example has had to do a bunch of gymnastics with the Brain class to take care of the interactions with VTK. I think we might end up having to do the same thing with mne-qt-browser.

The easiest way to get to the bottom of it is to first come up with some minimal example, which we more or less have with the command above. Then, if it's possible to make the PyQtGraphBrowser object do as little as possible by removing almost all functionality and attributes, that's the best first step. Hopefully that *doesn't* show the problem. If that's the case, then functionality / attributes / connections can be restored little by little to see where the problem is. Any problematic things can be dealt with in a cleanup that explicitly removes references, e.g.:

https://github.com/mne-tools/mne-python/blob/c87494462b7935e592efc0dcdc9553de1d0ccc5e/mne/viz/_brain/_brain.py#L720-L752

**Comment by GuillaumeFavelier** [comment]
Just adding that the `_clean()` function above is called in `closeEvent()`.

**Comment by larsoner** [comment]
@marsipu I'm working on this a little bit now. FYI you'll need to be on the branch from this PR to properly test:

https://github.com/sphinx-gallery/sphinx-gallery/pull/917

**Comment by larsoner** [comment]
Possible solution in https://github.com/mne-tools/mne-qt-browser/pull/61 -- @marsipu I'm out of time for the day, but if you merge that, feel free to push a commit here with `[circle full]` and we can see if it builds successfully (CircleCI should use latest `mne-qt-browser:main`)!

File: .circleci/config.yml [hunk 1]
```diff
@@ -110,7 +110,7 @@ jobs:
               echo "export MNE_3D_OPTION_ANTIALIAS=false" >> $BASH_ENV
               source tools/get_minimal_commands.sh
               echo "export MNE_3D_BACKEND=pyvistaqt" >> $BASH_ENV
-              echo "export MNE_BROWSER_BACKEND=matplotlib" >> $BASH_ENV
+              echo "export MNE_BROWSER_BACKEND=qt" >> $BASH_ENV
```
Comment (larsoner) [inline .circleci/config.yml:113]:
```suggestion
              echo "export MNE_BROWSER_BACKEND=qt" >> $BASH_ENV
               echo "export MNE_SKIP_INSTANCE_ASSERTIONS='brain,backgroundpotter,plotter,vtkpolydata,_renderer'" >> $BASH_ENV
```

File: tools/circleci_dependencies.sh [hunk 2]
```diff
@@ -23,8 +23,8 @@ else  # standard doc build
 	echo "Installing doc build dependencies"
 	python -m pip uninstall -y pydata-sphinx-theme
 	python -m pip install --upgrade --progress-bar off --only-binary "numpy,scipy,matplotlib,pandas,statsmodels" -r requirements.txt -r requirements_testing.txt -r requirements_doc.txt
-	python -m pip install --upgrade --progress-bar off https://github.com/mne-tools/mne-qt-browser/zipball/main
-	python -m pip install --progress-bar off https://github.com/sphinx-gallery/sphinx-gallery/zipball/master https://github.com/pyvista/pyvista/zipball/main https://github.com/pyvista/pyvistaqt/zipball/main
+	python -m pip install --upgrade --progress-bar off https://github.com/larsoner/mne-qt-browser/zipball/disconnect
+	python -m pip install --progress-bar off https://github.com/larsoner/sphinx-gallery/zipball/del https://github.com/pyvista/pyvista/zipball/main https://github.com/pyvista/pyvistaqt/zipball/main
```
Comment (larsoner) [inline tools/circleci_dependencies.sh:27]:
```suggestion
	python -m pip install --upgrade --progress-bar off https://github.com/larsoner/sphinx-gallery/zipball/del https://github.com/pyvista/pyvista/zipball/main https://github.com/pyvista/pyvistaqt/zipball/main
```

File: tools/circleci_dependencies.sh [hunk 3]
```diff
@@ -23,8 +23,8 @@ else  # standard doc build
 	echo "Installing doc build dependencies"
 	python -m pip uninstall -y pydata-sphinx-theme
 	python -m pip install --upgrade --progress-bar off --only-binary "numpy,scipy,matplotlib,pandas,statsmodels" -r requirements.txt -r requirements_testing.txt -r requirements_doc.txt
-	python -m pip install --upgrade --progress-bar off https://github.com/mne-tools/mne-qt-browser/zipball/main
-	python -m pip install --progress-bar off https://github.com/sphinx-gallery/sphinx-gallery/zipball/master https://github.com/pyvista/pyvista/zipball/main https://github.com/pyvista/pyvistaqt/zipball/main
+	python -m pip install --upgrade --progress-bar off https://github.com/larsoner/mne-qt-browser/zipball/disconnect
+	python -m pip install --upgrade --progress-bar off https://github.com/larsoner/sphinx-gallery/zipball/del https://github.com/pyvista/pyvista/zipball/main https://github.com/pyvista/pyvistaqt/zipball/main
```
Comment (larsoner) [inline tools/circleci_dependencies.sh:27]:
```suggestion
        python -m pip uninstall -y sphinx-gallery
	python -m pip install --upgrade --progress-bar off https://github.com/larsoner/sphinx-gallery/zipball/del https://github.com/pyvista/pyvista/zipball/main https://github.com/pyvista/pyvistaqt/zipball/main
```

File: .circleci/config.yml [hunk 4]
```diff
@@ -110,7 +110,8 @@ jobs:
...
+              echo "export MNE_SKIP_INSTANCE_ASSERTIONS='brain,backgroundplotter,plotter,vtkpolydata,_renderer'" >> $BASH_ENV
```
Comment (larsoner) [inline .circleci/config.yml:114]:
```suggestion
```

**Comment by larsoner** [comment]
Okay, the full build passed:

https://app.circleci.com/pipelines/github/mne-tools/mne-python/13033/workflows/3254276e-f4a5-4edd-a386-0f2bef2d8928/jobs/42014

Ready for review/merge from my end @agramfort !

**Comment by larsoner** [comment]
I'm actually going to put this in since @marsipu is happy already. That way, we can get things built and have tomorrow and Tues to look them over. But @agramfort or others if you have comments, feel free to make them here and I can make a follow-up PR!

**Pull request merged**